File: frontend/src/lib/api/scans.ts
import { request } from "./client";
import type { ScanCreate, ScanRead } from "@/lib/types/scan";
import type { AssetRead } from "@/lib/types/asset";

/**
 * GET /scans (optionally filtered by project_id) — used by the
 * /projects/[projectId] page (Task 12) to list a project's scans.
 */
export function listScans(projectId?: string): Promise<ScanRead[]> {
  const query = projectId ? `?project_id=${encodeURIComponent(projectId)}` : "";
  return request<ScanRead[]>(`/scans${query}`);
}

/** POST /scans — used by ScanCreateButton (Task 12). */
export function createScan(payload: ScanCreate): Promise<ScanRead> {
  return request<ScanRead>("/scans", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

/** GET /scans/{id} — used by the /projects/[projectId]/scans/[scanId] page (Phase 3 Task 6). */
export function getScan(scanId: string): Promise<ScanRead> {
  return request<ScanRead>(`/scans/${scanId}`);
}

/** GET /scans/{id}/assets — used by the /projects/[projectId]/scans/[scanId] page (Phase 3 Task 6). */
export function getScanAssets(scanId: string): Promise<AssetRead[]> {
  return request<AssetRead[]>(`/scans/${scanId}/assets`);
}

File: backend/app/api/routes/scans.py
import uuid

from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.api.deps import get_db
from app.models.asset import Asset
from app.models.scan import Scan
from app.schemas.asset import AssetRead
from app.schemas.scan import ScanCreate, ScanRead, ScanStatusRead
from app.services import scan_service

router = APIRouter(prefix="/scans", tags=["scans"])


@router.post("", response_model=ScanRead, status_code=status.HTTP_201_CREATED)
def create_scan(data: ScanCreate, db: Session = Depends(get_db)) -> Scan:
    return scan_service.create_scan(db, data)


@router.get("", response_model=list[ScanRead])
def list_scans(project_id: uuid.UUID | None = None, db: Session = Depends(get_db)) -> list[Scan]:
    return scan_service.list_scans(db, project_id)


@router.get("/{scan_id}", response_model=ScanRead)
def get_scan(scan_id: uuid.UUID, db: Session = Depends(get_db)) -> Scan:
    return scan_service.get_scan(db, scan_id)


@router.get("/{scan_id}/status", response_model=ScanStatusRead)
def get_scan_status(scan_id: uuid.UUID, db: Session = Depends(get_db)) -> Scan:
    return scan_service.get_scan(db, scan_id)


@router.post("/{scan_id}/cancel", response_model=ScanRead)
def cancel_scan(scan_id: uuid.UUID, db: Session = Depends(get_db)) -> Scan:
    return scan_service.cancel_scan(db, scan_id)


@router.get("/{scan_id}/assets", response_model=list[AssetRead])
def get_scan_assets(scan_id: uuid.UUID, db: Session = Depends(get_db)) -> list[Asset]:
    return scan_service.get_scan_assets(db, scan_id)
